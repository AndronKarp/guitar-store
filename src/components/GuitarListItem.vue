<template>
  <b-col>
    <router-link
      :to="{
        name: 'GuitarDescription',
        params: { slug: guitar.slug }
      }"
    >
      <b-card
        class="m-2"
        style="min-width: 280px;"
        :img-src="require(`../assets/${guitar.image}`)"
        :img-alt="guitar.model"
      >
        <b-card-title>{{ guitar.brand }} {{ guitar.model }}</b-card-title>
        <b-card-text>{{ guitar.price | currency }}</b-card-text>
      </b-card>
    </router-link>
  </b-col>
</template>

<script>
export default {
  props: {
    guitar: {
      type: Object,
      required: true
    }
  }
};
</script>

<style></style>
