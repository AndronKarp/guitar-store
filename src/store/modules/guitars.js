export default {
  state: {
    guitars: [],
    areGuitarsFetched: false
  },
  getters: {
    guitars(state) {
      return state.guitars;
    },
    areGuitarsFetched(state) {
      return state.areGuitarsFetched;
    }
  },
  mutations: {
    pushToGuitars(state, guitar) {
      state.guitars.push(guitar);
    },
    setNewGuitarQuantity(state, { guitar, value }) {
      guitar.quantity = value;
    },
    updateAreGuitarsFetchedStatus(state, { status }) {
      state.areGuitarsFetched = status;
    }
  },
  actions: {
    addToGuitars(store, guitar) {
      store.commit("pushToGuitars", guitar);
    },
    updateGuitarQuantity(context, { guitar, value }) {
      context.commit("setNewGuitarQuantity", { guitar, value });
    },
    setAreGuitarsFetchedStatusToTrue(store) {
      store.commit("updateAreGuitarsFetchedStatus", { status: true });
    }
  }
};
