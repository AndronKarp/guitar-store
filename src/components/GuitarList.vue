<template>
  <b-container fluid class="mt-3">
    <b-row cols-xl="4" cols-lg="3" cols-md="2" cols="1">
      <GuitarListItem
        v-for="(guitar, index) in guitars"
        :key="index"
        :guitar="guitar"
      />
    </b-row>
  </b-container>
</template>

<script>
import GuitarListItem from "./GuitarListItem";
import { mapGetters } from "vuex";

export default {
  computed: {
    ...mapGetters(["guitars"])
  },
  components: {
    GuitarListItem
  }
};
</script>

<style></style>
