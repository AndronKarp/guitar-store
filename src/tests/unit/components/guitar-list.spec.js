import { shallowMount, createLocalVue } from "@vue/test-utils";
import GuitarList from "@/components/GuitarList";
import Vuex from "vuex";
import guitars from "@/store/modules/guitars";
import BootstrapVue from "bootstrap-vue";
import GuitarListItem from "@/components/GuitarListItem";

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(BootstrapVue);

describe("GuitarList.vue", () => {
  let state;
  let store;
  let wrapper;

  beforeEach(() => {
    state = {
      guitars: [{ id: 0 }, { id: 1 }, { id: 2 }]
    };
    store = new Vuex.Store({
      modules: {
        guitars: {
          state,
          getters: guitars.getters
        }
      }
    });
    wrapper = shallowMount(GuitarList, { localVue, store });
  });
  test("renders list of guitars", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("accepts child GuitarListItem components", () => {
    const guitarListItems = wrapper.findAllComponents(GuitarListItem);
    expect(guitarListItems.exists()).toBe(true);
  });

  test("passes a binded guitar prop to GuitarListItem component", () => {
    const guitarListItem = wrapper.findComponent(GuitarListItem);
    expect(state.guitars).toContain(guitarListItem.vm.guitar);
  });
});
