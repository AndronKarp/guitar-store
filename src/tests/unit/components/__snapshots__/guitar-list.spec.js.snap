// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuitarList.vue renders list of guitars 1`] = `
<b-row-stub tag="div" cols="1" colsmd="2" colslg="3" colsxl="4">
  <guitarlistitem-stub guitar="[object Object]"></guitarlistitem-stub>
  <guitarlistitem-stub guitar="[object Object]"></guitarlistitem-stub>
  <guitarlistitem-stub guitar="[object Object]"></guitarlistitem-stub>
</b-row-stub>
`;
